adding info about running tests (#688)
Azure/azure-sdk-for-c: CONTRIBUTING.md

### Compiler Options

> By default, when building project with no options, next static libraries are generated:

Reviewer: The original comment had a typo, it should say "no static..."

Reviewer 2: I think the intent is to list the set of libraries that generated when no options are provided.

It isn't true that "no static libraries are generated". The libraries listed below (az_core, az_iot, etc.) do get generated.

Reviewer 2: ```suggestion
By default, when building the project with no options, the following static libraries are generated:
```

- ``Libraries``:
  - az_core
    - az_span, az_http, az_json, etc.
  - az_iot
    -  iot_provisioning, iot_hub, etc.
  - az_storage_blobs
    -  Storage SDK blobs client.
  - az_noplatform

>     - Library that provides a basic returning error for platform abstraction as AZ_NOT_IMPLEMENTED. This ensures project can be compiled without the need to provide any specific platform implementation. This is useful if you want to use az_core without platform specific functions likes Mutex or Time. 

Reviewer 2: A basic ...what?

>     - Library that provides a basic returning error for platform abstraction as AZ_NOT_IMPLEMENTED. This ensures project can be compiled without the need to provide any specific platform implementation. This is useful if you want to use az_core without platform specific functions likes Mutex or Time. 

Reviewer 2: ```suggestion
    - Library that provides a basic returning error for platform abstraction as AZ_NOT_IMPLEMENTED. This ensures the project can be compiled without the need to provide any specific platform implementation. This is useful if you want to use az_core without platform specific functions like `mutex` or `time`. 
```

    -  Storage SDK blobs client.
  - az_noplatform
    - Library that provides a basic returning error for platform abstraction as AZ_NOT_IMPLEMENTED. This ensures project can be compiled without the need to provide any specific platform implementation. This is useful if you want to use az_core without platform specific functions likes Mutex or Time. 
  - az_nohttp

>   - az_posix (on Lin/Mac)

Reviewer: Why are these dropped but az_notplatform stays?

Author: because az_notttp is generated/compiled with default compiler-options.  
az_posix and az_win32 are mentioned below with the required compiler-option

>     -  Library that provides a basic returning error when calling HTTP stack. Similar to az_noplatform, this library ensures project can be compiled without requiring any HTTP stack implementation. This us useful if you want to use az_core without az_http functionality.

Reviewer 2: ```suggestion
    -  Library that provides a basic returning error when calling HTTP stack. Similar to az_noplatform, this library ensures the project can be compiled without requiring any HTTP stack implementation. This is useful if you want to use `az_core` without `az_http` functionality.
```

> The next compiler options are available for adding/removing project features.

Reviewer 2: ```suggestion
The following compiler options are available for adding/removing project features.
```

<table>
<tr>
<td>Option</td>
<td>Description</td>
<td>Default Value</td>
</tr>
<tr>
<td>UNIT_TESTING</td>
<td>Generates Unit Test for compilation. When turning this option ON, cmocka is a required dependency for compilation.<br>After Compiling, use `ctest` to run Unit Test.</td>
<td>OFF</td>
</tr>
<tr>
<td>UNIT_TESTING_MOCK_ENABLED</td>
<td>This option works only with GCC. It uses -ld option from linker to mock functions during unit test. This is used to test platform or HTTP functions by mocking the return values.</td>
<td>OFF</td>
</tr>
<tr>
<td>BUILD_PRECONDITIONS</td>

> <td>Turning this option ON would remove all method contracts. This us typically for shipping libraries for production to make it as much optimized as possible.</td>

Reviewer 2: ```suggestion
<td>Turning this option ON would remove all method contracts. This is typically for shipping libraries for production to make it as much optimized as possible.</td>
```

<td>ON</td>
</tr>
<tr>
<td>BUILD_CURL_TRANSPORT</td>
<td>This option requires Libcurl dependency to be available. It generates an HTTP stack with libcurl for az_http to be able to send requests thru the wire. This library would replace the no_http.</td>
<td>OFF</td>
</tr>
<tr>
<td>BUILD_PAHO_TRANSPORT</td>
<td>This option requires paho-mqtt dependency to be available. Provides Paho MQTT support for iot.</td>
<td>OFF</td>
</tr>
<tr>
<td>AZ_PLATFORM_IMPL</td>
<td>This option can be set to any of the next values:<br>- No_value: default value is used and no_platform library is used.<br>- "POSIX": Provides implementation for Linux and Mac systems.<br>- "WIN32": Provides platform implementation for Windows based system<br>- "USER": Tells cmake to use an specific implementation provided by user. When setting this option, user must provide an implementation library and set option `AZ_USER_PLATFORM_IMPL_NAME` with the name of the library (i.e. <code>-DAZ_PLATFORM_IMPL=USER -DAZ_USER_PLATFORM_IMPL_NAME=user_platform_lib</code>). cmake will look for this library to link az_core</td>
<td>No_value</td>
</tr>
</table>

> - ``Samples``: Whenever UNIT_TESTING is ON, samples are built using the default PAL (see [running samples section](#running-samples)). This means that running samples would throw errors like:

Reviewer 2: This document mentions it uses the default PAL, but the error when running the sample talks about `no_op HTTP implementation`. Does it need a custom PAL as well or is only the HTTP stack missing?

i.e. cmake -DBUILD_CURL_TRANSPORT=ON ..
```

When running cmake, next options can be used to change the output libraries/Pal/Samples:

- `BUILD_CURL_TRANSPORT`: This option would build an HTTP transport library using CURL. It requires libcurl to be installed (vcpkg or globally). This option will make samples to be linked with this HTTP and be functional to send HTTP requests.
## Running Tests and Samples

```bash
cmake -DBUILD_CURL_TRANSPORT=ON ..
cmake --build .
```
### Unit tests

> - `UNIT_TESTING`: This option requires cmocka to be installed and it will generate unit tests for each project.

Reviewer: Add a reference to the cmocka requirement

Author: it is mentioned in the #compiler-options section as:
><td>UNIT_TESTING</td>
<td>Generates Unit Test for compilation. When turning this option ON, cmocka is a required dependency for compilation.<br>After Compiling, use `ctest` to run Unit Test.</td>

See [compiler options section](#compiler-options) to learn about how to build samples with HTTP implementation in order to be runnable.

> After building samples with HTTP stack, set next environment variables for enabling credentials. Samples will read these values from env and use it to log in to Azure Service like Storage or KeyVault. Learn about the supported authentication [client secret here](https://docs.microsoft.com/en-us/azure/active-directory/azuread-dev/v1-oauth2-on-behalf-of-flow#service-to-service-access-token-request).

Reviewer: ```suggestion
After building samples with HTTP stack, set the environment variables for credentials.  The samples read these environment values to authenticate to Azure services.  See [client secret here](https://docs.microsoft.com/en-us/azure/active-directory/azuread-dev/v1-oauth2-on-behalf-of-flow#service-to-service-access-token-request) for additional details on Azure authentication.
```

Reviewer 2: @gilbertw, I edited your suggestion to remove the extra white space after the period in `... to Azure services.  See [client ...` (assuming we use single white space after periods in most places). 

Please undo if I am mistaken.

```bash
# On linux, set env var like this. For Windows, do it from advanced settings/ env variables

# KEY-VAULT Sample
export AZURE_TENANT_ID="????????-????-????-????-????????????"
export AZURE_CLIENT_ID="????????-????-????-????-????????????"
export AZURE_CLIENT_SECRET="????????????"

> export AZURE_KEYVAULT_URL="https://????.????.azure.net"

Reviewer: ```suggestion
export AZURE_KEYVAULT_URL="https://???????????.??"
```

Reviewer: In a private cloud the url won't be .azure.net

# STORAGE Sample (only 1 env var required)

> # URL must contain a valid container, blob and SaS token
> export AZURE_STORAGE_URL="https://storageAccount.blob.core.windows.net/container/blob?sv=xxx&ss=xx&srt=xx&sp=xx&se=xx&st=xxx&spr=https,http&sig=xxx"

Reviewer: ```suggestion
# URL must contain a valid container, blob and SaS token
# e.g "https://storageAccount.blob.core.windows.net/container/blob?sv=xxx&ss=xx&srt=xx&sp=xx&se=xx&st=xxx&spr=https,http&sig=xxx"
export AZURE_STORAGE_URL="https://??????????????"
```
